packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,507 @@
+      duration: kThemeAnimationDuration
+    );
+
+    _gapControllers[index] = gapController;
+
+    final CurvedAnimation gapAnimation = new CurvedAnimation(
+      parent: gapController,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+
+    _gapAnimations[index] = gapAnimation;
+
+    gapAnimation.addListener(_handleTick);
+  }
+
+  void _insertNulls(int index) {
+    _startControllers.insert(index, null);
+    _endControllers.insert(index, null);
+    _gapControllers.insert(index, null);
+    _startAnimations.insert(index, null);
+    _endAnimations.insert(index, null);
+    _gapAnimations.insert(index, null);
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _startControllers = new List<AnimationController>();
+    _endControllers = new List<AnimationController>();
+    _gapControllers = new List<AnimationController>();
+    _startAnimations = new List<CurvedAnimation>();
+    _endAnimations = new List<CurvedAnimation>();
+    _gapAnimations = new List<CurvedAnimation>();
+
+    for (int i = 0; i < children.length; i += 1) {
+      _insertNulls(i);
+
+      if (children[i].isGap) {
+        _initGap(i);
+        _gapControllers[i].value = 1.0; // Gaps must be wide when first built.
+      } else {
+        _initSlice(i);
+      }
+    }
+    _assertGaps(children);
+    _fixSliceStates(children);
+  }
+
+  @override
+  void dispose() {
+    for (AnimationController controller in _startControllers)
+      controller?.dispose();
+    for (AnimationController controller in _endControllers)
+      controller?.dispose();
+    for (AnimationController controller in _gapControllers)
+      controller?.dispose();
+    super.dispose();
+  }
+
+  void _handleTick() {
+    setState(() {
+      // The animation's state is our build state, and it changed already.
+    });
+  }
+
+  bool _consecutiveGaps(List<MergeableMaterialItem> children) {
+    for (int i = 0; i < config.children.length - 1; i += 1) {
+      if (config.children[i].isGap && config.children[i + 1].isGap) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void _assertGaps(List<MergeableMaterialItem> children) {
+    // Check for consecutive gaps.
+    assert(!_consecutiveGaps(children));
+
+    // First and last children must not be gaps.
+    if (children.isNotEmpty) {
+      assert(!children.first.isGap);
+      assert(!children.last.isGap);
+    }
+  }
+
+  void _fixTwoSliceStates(int i, int j, double value) {
+    if (!_endControllers[i].isAnimating && !_startControllers[j].isAnimating) {
+      _endControllers[i].value = value;
+      _startControllers[j].value = value;
+    }
+  }
+
+  void _fixSliceStates(List<MergeableMaterialItem> children) {
+    // First and last slices must be round.
+    if (children.isNotEmpty) {
+      _startControllers[0].value = 1.0;
+      _endControllers[children.length - 1].value = 1.0;
+    }
+
+    for (int i = 1; i < children.length; i++) {
```

Comment:
use `i += 1` instead of `i++`. In some languages, with some compilers, `i++` is suboptimal, so best practice is to avoid it.